File: Leetcode 322 - coin-change/322.coin-change.cpp
/*
 * @lc app=leetcode id=322 lang=cpp
 *
 * [322] Coin Change
 */

// @lc code=start
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int coinChange(vector<int> &coins, int amount) {
    vector<int> dp(amount + 1, amount + 1);
    dp[0] = 0;

    for (int i = 1; i <= amount; i++) {
      for (int coin : coins) {
        if (coin <= i) {
          dp[i] = min(dp[i], dp[i - coin] + 1);
        }
      }
    }

    return dp[amount] > amount ? -1 : dp[amount];
  }
};

int main() {
  Solution sol;

  vector<int> coins1 = {1, 2, 5};
  cout << sol.coinChange(coins1, 11) << endl; // 3

  vector<int> coins2 = {2};
  cout << sol.coinChange(coins2, 3) << endl; // -1

  vector<int> coins3 = {1};
  cout << sol.coinChange(coins3, 0) << endl; // 0

  vector<int> coins4 = {186, 419, 83, 408};
  cout << sol.coinChange(coins4, 6249) << endl; // 20
}
// @lc code=end

File: Leetcode 322 - coin-change/322.coin-change.java
/*
 * @lc app=leetcode id=322 lang=java
 *
 * [322] Coin Change
 */

// @lc code=start

import java.util.Arrays;

class Solution {

  public int coinChange(int[] coins, int amount) {
    int[] dp = new int[amount + 1];
    Arrays.fill(dp, amount + 1);
    dp[0] = 0;

    // Bottom-up dynamic programming
    for (int i = 1; i <= amount; i++) {
      for (int coin : coins) {
        if (coin <= i) {
          dp[i] = Math.min(dp[i], dp[i - coin] + 1);
        }
      }
    }

    return dp[amount] > amount ? -1 : dp[amount];
  }

  // Time Complexity: O(n * amount)
  // Space Complexity: O(amount)

  public static void main(String[] args) {
    Solution solution = new Solution();

    int[] coins = { 1, 2, 5 };
    int amount = 11;
    System.out.println(solution.coinChange(coins, amount)); // 3

    coins = new int[] { 2 };
    amount = 3;
    System.out.println(solution.coinChange(coins, amount)); // -1

    coins = new int[] { 1 };
    amount = 0;
    System.out.println(solution.coinChange(coins, amount)); // 0
  }
}
// @lc code=end
